pkg/provisioner/controller.go

```diff
@@ -291,6 +293,103 @@ func (c *obcController) handleProvisionClaim(key string, obc *v1alpha1.ObjectBuc
 	// to control access to static buckets via RBAC rules on storage classes.
 	isDynamicProvisioning := isNewBucketByStorageClass(class)
 
+	// Re-Get the claim in order to shorten the race condition where the claim was deleted after provisioning started
+	obc, err = claimForKey(key, c.libClientset)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return fmt.Errorf("OBC was lost before we could provision: %v", err)
+		}
+		return err
+	}
+
+	// Older versions of this library set obc.Spec.BucketName with the name of the bucket generated
+	// from obc.Spec.GenerateBucketName, which causes an error condition when updating to a newer
+	// version of this library if provisioning runs again. If the bucket name has the
+	// GenerateBucketName prefix, we can assume that this was originally intended to be a generated
+	// name.
+	if obc.Spec.GenerateBucketName != "" && obc.Spec.BucketName != "" {
```

Comment:
The design doc allows both of these names to be specified and states in that case `bucketName` has precedence over `generateBucketName`. So, unless you have API validation or an admission controller to remove `bucketName`, when `generateBucketName` is specified, this 'if` will blank out bucket name and thus change the hierarchy described in the doc. This is a backwards incompatible change, if I understand it correctly.

Reply:
It seems like the code as implemented and the design doc are out of sync. See the code snippet below which I have not modified.

```go
func composeBucketName(obc *v1alpha1.ObjectBucketClaim) (string, error) {
	if obc.Spec.BucketName == "" && obc.Spec.GenerateBucketName == "" {
		return "", fmt.Errorf("expected either bucketName or generateBucketName defined")
	}
	if obc.Spec.BucketName != "" && obc.Spec.GenerateBucketName != "" {
		return "", fmt.Errorf("cannot define both bucketName and generateBucketName")
	}
	bucketName := obc.Spec.BucketName
	if bucketName == "" {
		bucketName = generateBucketName(obc.Spec.GenerateBucketName)
	}
	return bucketName, nil
}
```

GitLens in my editor shows that the line of behavior I have been taking to be canon was added by you.
```go
	if obc.Spec.BucketName != "" && obc.Spec.GenerateBucketName != "" {
		return "", fmt.Errorf("cannot define both bucketName and generateBucketName")
	}
```
```
allow obc to have no bucket related fields for "brownfield" (#88)

* change ob ClaimRef from UID to ObjectReference

Signed-off-by: jeffvance <jeff.h.vance@gmail.com>

* added UID to ob.ClaimRef

Signed-off-by: jeffvance <jeff.h.vance@gmail.com>

* allow obc to have no bkt name fields for brownfield

Signed-off-by: jeffvance <jeff.h.vance@gmail.com>
```

Should I instead drop this recovery code and change the behavior to prefer `bucketName` over `generateBucketName`?

Reply:
That's a good question. The code is not necessarily wrong but for sure the code and doc are not in sync here. I lean toward creating a small issue against the doc and leaving the code as is.

Reply:
If the code gives precedence to `bucketName` as in the doc, that makes the upgrade recovery case easier. We would no longer have to infer that a `bucketName` that has `generateBucketName` as a prefix should originally have been only `generateBucketName`. I think maybe this is a slightly better option.